### ISSUE TITLE
self.__class__() not supported

### ISSUE THREAD

**Issue by timabbott** [python/mypy#1195]
Mypy is OK with accessing `self.__class__` but not `self.__class__(...)`

Reproducer:
```
class S(object):
    def foo(self):
        return self.__class__()
```
Error:
```
/home/tabbott/foo.py: note: In member "foo" of class "S":
/home/tabbott/foo.py:4: error: "type" not callable
```

**Comment by gvanrossum** [comment]
Well, maybe there should be a better error message, but calling
self.__class__() to construct an instance is very problematic. A
subclass constructor might have a different signature that is
unknowable when you are checking the base class. So unless in your
world you have a strong rule that subclass constructors must conform
to a certain signature, this should be flagged IMO. (OTOH a way to
declare that you do have such a rule might be useful -- maybe as an
annotation on __init__()?)

**Comment by JukkaL** [comment]
Yeah, that's the reason why this doesn't work right now. Having an `__init__` annotation would be a reasonable way to getting to work. Maybe provide a decorator in `typing` that could be used for this:

```
from typing import whatever

class A:
    @whatever
    def __init__(self, x: int) -> None: ...

class B(A):
    def __init__(self) -> None: ...  # error!
```
Not sure what we would call it, though.

**Comment by gvanrossum** [comment]
This no longer produces an error.
